ES|QL date_range data type

shmuelhanoch opened the pull request:
Basic support for DATE_RANGE field type.

Commit:
Basic support for DATE_RANGE field type in ESQL

nik9000 commented on .idea/runConfigurations/Debug_Elasticsearch__node_2_.xml:
```diff
@@ -6,6 +6,10 @@
+    <RunnerSettings RunnerId="Debug">
+      <option name="DEBUG_PORT" value="5008" />
+      <option name="LOCAL" value="false" />
+    </RunnerSettings>
```
I don't know what makes these! I've made them before on accident. Probably best to revert it. If you can figure out how you made it then we'll see how to include it. Or how to not make it. Or something.

nik9000 commented on .gitignore:
```diff
@@ -76,3 +76,4 @@ server/src/main/resources/transport/definitions/manifest.txt
 
 # JEnv
 .java-version
+.aider*
```
Probably worth adding this in another PR that explain what it's for.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -420,10 +514,14 @@ protected void parseCreateField(DocumentParserContext context) throws IOExceptio
+            case RangeType.IP -> parseIpRangeFromCidr(parser);
+            case RangeType.DATE -> parseDateRange(parser.text());
+            default -> parseGeneralRange(parser, start);
+        };
```
Worth reverting this bit I think.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/index/mapper/TestBlock.java:
```diff
@@ -582,4 +588,61 @@ public void close() {
+
+        @Override
+        public void close() {
+
```
You should `Releasables.close(from, to)` here - that *might* be your leak.

nik9000 commented:
I'd add another test to `x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/10_basic.yml` that loads the range and asserts that output looks right. That should work given when you are, but it's really simple to point at.

After than, I'd try and get the rest of the tests passing.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -349,6 +351,97 @@ public Query rangeQuery(
+            public AllReader reader(LeafReaderContext context) throws IOException {
+                var docValues = context.reader().getBinaryDocValues(fieldName);
+                return new DateRangeDocValuesReader(docValues);
+                /*
```
We probably shouldn't commit this without at least an explanation. We can just zap it for now.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -349,6 +351,97 @@ public Query rangeQuery(
+
+        @Override
+        protected int docId() {
+            return 0;
```
This'll need implementing. Without it we'll incorrectly try and reuse this when loading after a topn.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -349,6 +351,97 @@ public Query rangeQuery(
+
+        @Override
+        public String toString() {
+            return "";
```
It'd be nice to have something readable here.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -481,6 +481,32 @@ public final AggregateMetricDoubleBlock newAggregateMetricDoubleBlock(
+                }
+            }
+            return builder.build();
+        }
```
I think we'd be better off doing a `DateRangeBlock` that contains to constant blocks.

nik9000 reviewed (commented):
Makes sense to me. I think it's time to drop draft from this, get all the tests happy, and get it in. Then we can extend it with stuff like supporting this field type in top n (not as a sort key, but as a *sorted* field), COALESCE, TO_STRING, etc. And maybe a `TO_DATE_RANGE` function too. Lots of separate stuff we can build on top of this.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeBlock.java:
```diff
@@ -0,0 +1,73 @@
+    static boolean equals(DateRangeBlock lhs, DateRangeBlock rhs) {
+        if (lhs == rhs) return true;
+        if (lhs.getPositionCount() != rhs.getPositionCount()) return false;
+        return LongBlock.equals(lhs.getFromBlock(), rhs.getFromBlock()) && LongBlock.equals(lhs.getToBlock(), rhs.getToBlock());
```
Checkstyle is going to be upset about those `if` statements. This is java! It's like a cruise ship. Very wide.

nik9000 reviewed (commented):
Looks great! Let's get this un-drafted. I'm pretty comfortable getting this in if CI is happy.

Once CI is truly happy I'll give it one more look, hopefully just to find follow up stuff.

Commit:
[CI] Update transport version definitions

shmuelhanoch committed:
Update docs/changelog/133309.yaml

Commit:
Suppres two more tests

Commit:
Two more tests fixes

Commit:
Two more tests

Commit:
Some more infra for new tests

Commit:
Some more fixes for tests

Commit:
And some more tests

Commit:
And some more tests

Commit:
Fix capanility check in CSV tests and some more fixes

Commit:
Block lookup join

Commit:
Block lookup join

Commit:
Relaxed the previous transport version check, which was in fact always false - replace it with something that can work for now.

Commit:
Fixed a typo

Commit:
Style fixes

alex-spies reviewed (changes requested):
It looks like `DATE_RANGE` is not tested in the `AllSupprtedFieldsTestCase`. Can we please double check this? Without testing `FROM * | KEEP *`-like queries with this type, we could cause major breakage when merging this.

Commit:
More fixes: block lookup join. Delete date_range from unsupported yaml test, it's not unsupported anymore and there is a normal yaml test for it.

Commit:
Supress RestEsqlIT because complex queries like this are not supported yet

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/data/decades.csv:
```diff
@@ -1,14 +1,14 @@
-2000-01-01..2010-01-01,           2000, Innovation Explosion Decade
-2010-01-01..2020-01-01,           2010, Renaissance Decade
-2020-01-01..2030-01-01,           2020, Empowerment Era
+date_range:date_range, decade:integer, description:keyword
```
Why was this reformatted?

shmuelhanoch replied:
Fixed

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -2028,3 +2028,23 @@ from employees
+
+dateRange
+required_capability: date_range_field_type
+from date_ranges;
```
Convention is to capitalise `FROM`

craigtaverner commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:
```diff
@@ -166,56 +162,54 @@ unsupported:
   - match: { columns.3.name: date_nanos }
   - match: { columns.3.type: date_nanos }
-  - match: { columns.4.name: date_range }
+  - match: { columns.4.name: dense_vector }
```
Normally what we've done in this file is simply update the type to now be supported, as opposed to removing it from the file. This makes for a much simpler change. If you really want to remove it from the file, then consider adding additional tests to the 30_types.yml file to test the new type. As it currently stands, you no longer have date_range mentioned in any yaml tests, which is not so good.

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -85,10 +86,10 @@ public static TypeResolution isRepresentableExceptCountersDenseVectorAndAggregat
         return isType(
             e,
-            dt -> isRepresentable(dt) && dt != DENSE_VECTOR && dt != AGGREGATE_METRIC_DOUBLE,
+            dt -> isRepresentable(dt) && dt != DENSE_VECTOR && dt != AGGREGATE_METRIC_DOUBLE && dt != DATE_RANGE,
```
This seems to exclude this type from many functions. Is that because we plan to add that support in followup PRs?

nik9000 replied:
Probably worth a comment that this is temporary while under construction.

I'm not 100% sure why we need this to be honest though - I had thought we didn't try sending these "under construction" types in tests already.

craigtaverner commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/type/DataType.java:
```diff
@@ -270,6 +271,10 @@ public enum DataType implements Writeable {
      */
     DATE_NANOS(builder().esType("date_nanos").estimatedSize(Long.BYTES).docValues().supportedOnAllNodes()),
+    /**
+     * Represents a half-inclusive range between two dates.
```
Is it half-inclusive? I saw a comment in the RangeFieldMapper that indicates this is fully inclusive.

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeBlock.java:
```diff
@@ -0,0 +1,77 @@
+import org.elasticsearch.core.ReleasableIterator;
+
+/**
+ * Block that stores aggregate_metric_double values.
```
Really? I don't think so.

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeBlock.java:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Block that stores aggregate_metric_double values.
+ */
+public sealed interface DateRangeBlock extends Block permits DateRangeArrayBlock, ConstantNullBlock {
```
I don't see anything in this block that is Date or DateRange specific. It is basically a block of two long blocks. Should it not be called something more generic, like LongLongBlock, or TwoLongsBlock?

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeArrayBlock.java:
```diff
@@ -0,0 +1,218 @@
+
+import java.io.IOException;
+
+public final class DateRangeArrayBlock extends AbstractNonThreadSafeRefCounted implements DateRangeBlock {
```
As with DateRangeBlock, there is nothing in this class specific to Dates or DateRanges. How about `LongRangeArrayBlock`?

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeBlockBuilder.java:
```diff
@@ -0,0 +1,187 @@
+        return toBuilder;
+    }
+
+    public record DateRangeLiteral(Long from, Long to) implements GenericNamedWriteable {
```
The word `Literal` in the class name makes me think this is generated by the ES|QL parser, but as far as I can tell it is not used for that at all. It is just a container for a single date-range value. Perhaps the class should be `DateRange` or `DateRangeValue`?

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/DateRangeBlockBuilder.java:
```diff
@@ -0,0 +1,187 @@
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            return ESQL_DATE_RANGE_CREATED_VERSION;
```
If we rename this record to `LongRangeValue`, we could also rename this transport version to something more generic like `ESQL_LONG_RANGES`.

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ElementType.java:
```diff
@@ -64,7 +64,9 @@ public enum ElementType {
-    );
+    ),
+
+    DATE_RANGE(11, "DateRange", BlockFactory::newDateRangeBlockBuilder, DateRangeArrayBlock::readFrom);
```
How about `LONG_RANGE`?

craigtaverner reviewed (changes requested):
I got about half way through. I have many suggestions and questions. But my main concern was when I brought back the deleted yaml tests, I saw unexpected failures (date_range was `unsupported` with LIMIT 0, and without LIMIT 0, I got a null result, instead of a valid result, when sorting on another field). I'll push the tests in a commit, so they can be looked at. And I'll try continue the review tomorrow.

craigtaverner committed:
Removed unused field

craigtaverner committed:
Bring back deleted yaml tests

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/BlockFactory.java:
```diff
@@ -481,6 +481,34 @@ public final AggregateMetricDoubleBlock newAggregateMetricDoubleBlock(
+        }
+    }
+
+    public DateRangeBlock newDateRangeBlock(long[] fromValues, long[] toValues, int positions) {
```
This method appears entirely unused

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateRange.java:
```diff
@@ -0,0 +1,74 @@
+    );
+
+    private static final Map<DataType, BuildFactory> EVALUATORS = Map.ofEntries(
+        // TODO
```
OK. This looks missing!

craigtaverner replied:
The obvious case would be to support strings, which would simply mean to parse the date-range using the `..` literal syntax. You could have a csv-spec test like:

```
ROW text = "2025-10-10T12:00:00..2025-10-10T12:00:00"
EVAL date_range = TO_DATE_RANGE(text)
```

This would be sufficient to test this. The implementation would also be trivial, just add one `@ConvertEvaluator` method. I could push a small commit showing this, if you want?

craigtaverner replied:
I realize I can think of a compromise here, where we implement only the identity function, and leave the parse-string for a followup PR.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDateRange.java:
```diff
@@ -0,0 +1,74 @@
+        // TODO
+    );
+
+    @FunctionInfo(returnType = "DateRange", description = "TBD")
```
Add docs

craigtaverner commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockBuilderTests.java:
```diff
@@ -183,7 +183,7 @@ public void testCrankyConstantBlock() {
                     builder.copyFrom(random.block(), 0, random.block().getPositionCount());
                     try (Block built = builder.build()) {
-                        if (built instanceof AggregateMetricDoubleArrayBlock == false) {
+                        if (built instanceof AggregateMetricDoubleArrayBlock == false && built instanceof DateRangeArrayBlock == false) {
```
Does date_range really not support multi-values?

shmuelhanoch replied:
I didn't implemented `asVector` because it wasn't implemented in AMD. Do we need `asVector` for date_range? I'm not sure in which scenarios it's usually used

craigtaverner committed:
Refactor DateRange Block type (and element type) to LongRange

shmuelhanoch commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LongRangeBlock.java:
```diff
@@ -0,0 +1,77 @@
+     * Returns {@code true} if the given blocks are equal to each other, otherwise {@code false}.
+     * Two blocks are considered equal if they have the same position count, and contain the same
+     * values (including absent null values) in the same order. This definition ensures that the
+     * equals method works properly across different implementations of the AggregateMetricDoubleBlock interface.
```
Apart from the copy-paste accident, is there any reason why this implementation is not correct? I made it to match the same logic and same way that AggregateMetricDouble was to pass some tests, that failed without it (maybe ExtractorTest or something like that)

shmuelhanoch commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LongRangeBlockBuilder.java:
```diff
@@ -0,0 +1,187 @@
+        );
+
+        public DateRangeLiteral(StreamInput in) throws IOException {
+            this(in.readOptionalLong(), in.readOptionalLong());
```
Definitely no, I blindly copied from AggregateMetricDouble. Without having optional / nullable parts of the range I failed some tests like ExtractorTest, but maybe it's better to avoid nullable fields (because it's meaningless and shouldn't be representible) and fix the test itself...I can try to do that

nik9000 replied:
Yeah, I think it should be `long, long` - neither optional. We tend to represent unbounded ranges with `Long.MAX_VALUE`

craigtaverner replied:
I'm not saying we must do this, but it is worth thinking about. Ideally we would like a situation where adding new data types does not requiring adding new blocks or element types. And perhaps all 128-bit types could be backed by a single element type and block type. Wishful thinking? 😎

shmuelhanoch commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -61,6 +64,8 @@ public class RangeFieldMapper extends FieldMapper {
     public static final boolean DEFAULT_INCLUDE_UPPER = true;
     public static final boolean DEFAULT_INCLUDE_LOWER = true;
 
+    public static final TransportVersion ESQL_DATE_RANGE_CREATED_VERSION = TransportVersion.fromName("esql_date_range_created_version");
```
This place to define it was suggested by Nik as the only (I think) common subproject between `core-esql` (in which `DataType` is defined) and the place `BlockBuilder` is defined in. (`compute.data`). It was a simpler alterantice to add a new dependency.

 The CREATED_VERSION is copied again from `AggregateMetricDouble`: unlike the later, we have only one transport version related to `date_range`, so it obviously redundant, I'lll delete it

shmuelhanoch commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -342,6 +347,96 @@ public Query rangeQuery(
+            if (hasDocValues()) {
+                return new DateRangeDocValuesLoader(name());
+            }
+            throw new IllegalStateException("Cannot load blocks without doc values");
```
Maybe? But as suggested by Nik, since  doc value is the important one, to minimize the PR it's better to start from it.

nik9000 replied:
Yeah, in a separate PR is good.

shmuelhanoch commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/data/ages.csv:
```diff
@@ -1,11 +1,11 @@
-                60..80 , Senior
-                80..100, Elderly
-               100..200, Incredible
+0..2, Baby
```
I saw this space which seemed redundant and thought it's a good opportunity to fix it, if it's too irrelevant I'll get rid of this change

shmuelhanoch replied:
Fixed

nik9000 commented on x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java:
```diff
@@ -99,10 +100,14 @@ public static TypeResolution isRepresentableExceptCountersSpatialDenseVectorAndA
+                && DataType.isRepresentable(t)
+                && t != DENSE_VECTOR
+                && t != AGGREGATE_METRIC_DOUBLE
+                && t != DATE_RANGE,
```
:+1:

shmuelhanoch replied:
Changed to `isRepresentableExceptUnsortableTypes`, unless there is a better suggestion

nik9000 commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/AllSupportedFieldsTestCase.java:
```diff
@@ -461,7 +472,9 @@ private static boolean supportedInIndex(DataType t) {
-                CARTESIAN_POINT, CARTESIAN_SHAPE -> false;
+                CARTESIAN_POINT, CARTESIAN_SHAPE,
+                // Excluded for now, until a proper workaround
+                DATE_RANGE -> false;
```
This test only runs when you enable the release flag right now - and there `date_range` fields are all `null` because the support is snapshot only.

nik9000 commented on x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/AllSupportedFieldsTestCase.java:
```diff
@@ -438,6 +444,11 @@ private Matcher<?> expectedValue(DataType type) {
+            case DATE_RANGE -> {
+                // Currently, we cannot tell if all nodes support it or not so we treat it as unsupported.
+                // TODO: Fix this once we know the node versions.
+                yield nullValue();
```
For `DATE_RANGE` this should be `nullValue()` with the comment something like "date_range is unreleased".

shmuelhanoch commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -342,6 +347,96 @@ public Query rangeQuery(
+
+        @Override
+        public BlockLoader blockLoader(BlockLoaderContext blContext) {
+            if (hasDocValues()) {
```
It seems that in `AggregateMetricDouble` case, there is no such check. And there is also no check for `hasDockValues` (not sure where I got that from). 

I'm not sure about either of those checks, which of them is essential if any?

shmuelhanoch commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LongRangeBlock.java:
```diff
@@ -0,0 +1,77 @@
+        return b.isNull(firstValueIdx + valueIndex) ? -1 : Long.hashCode(b.getLong(firstValueIdx + valueIndex));
+    }
+
+    // TODO: those DateRange-specific sub-block getters, together with the AggregateMetricBuilder specific getters
```
Not out of date. Regardless the thing you referred to, and assuming we do have two different blocks, one for LongRange and one for AggregateMetricDouble, I would still consider this refactoring:

- `ConstantNullBlock` implements both `LongRangeBlock` and `AggregateMetricDoubleBlock` (and more "composite blocks" to come, maybe...).

- ATM, each of them has its own getter methods for the sub-blocks: `AggregateMetricDouble` has its 5 named getters, and `DateRangeBlock` has its own `fromBlock / toBlock` special getters. It means that ConstantNullBlock needs to support the union of those 7 methods, and it will need to add more if we'll add, say , "int range". It's somewhat silly because all of those getters does a similar thing, just "get the nth block"!

- The suggestion was to create a very small interface, "CompositeLikeBlock" or something, which has a single method: getNthBlock. All "composite blocks" like AMD or LongRange will implement it, and then ConstantNullBlock will only need a single method instead of <Sum of all numbers of all getters all composite blocks have).

Commit:
Review fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
small fixes

craigtaverner committed:
Added identity option to TO_DATE_RANGE, plus more tests

craigtaverner committed:
Merge branch 'main' into esql_date_range_data_type

theletterf committed:
Fix irregular spaces (#137014)

* Fix irregular spaces

* Update analysis-keyword-repeat-tokenfilter.md

* Update search-suggesters.md

* Update search-profile.md

DaveCTurner committed:
Test utility for `POST _features/_reset` (#137133)

We call `POST _features/_reset` in various places in the tests, only
sometimes asserting that the response is `200 OK`. This commit extracts
a utility that makes it harder to miss this assertion. It also adds the
`?error_trace` parameter so that on failure the details are visible in
the response, and `?master_timeout=-1` to avoid spurious timeouts if CI
is running a little slow.

DaveCTurner committed:
Revert "Test utility for `POST _features/_reset` (#137133)"

This reverts commit df67e2733f365778ab2dbbac2a8d418c55964af9.

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.qa.multi_node.EsqlSpecIT test {csv-spec:lookup-join-expression.LookupJoinExpressionWithTerm} #137157

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.qa.single_node.EsqlSpecIT test {csv-spec:lookup-join-expression.LookupJoinExpressionWithTerm} #137157

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.qa.single_node.GenerativeForkIT test {csv-spec:lookup-join-expression.LookupJoinExpressionWithTerm} #137160

elasticsearchmachine committed:
Mute org.elasticsearch.cluster.routing.allocation.decider.WriteLoadConstraintDeciderIT testCanRemainNotPreferredIsIgnoredWhenAllOtherNodesReturnNotPreferred #137162

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

dnhatn committed:
Try bulk load sorted ordinals across ranges (#137076)

This change implements the TODO for loading sorted ordinals in the TSDB 
codec. With ordinal range encoding, we can bulk-append the entire range 
for the next ordinal, rather than reading and appending one document at
a time.

Commit:
Fix serialization assymetry (writeOptional vs plain read), fixed VerifierTett

Commit:
Relax (again) the check in blockLoader in RangeFieldMapper, since it seem to fail in practice

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Auto commit changes from spotless

mosche committed:
Fix APM tracer to respect recording status of spans (#137015)

Co-authored-by: Jack Shirazi <jack.shirazi@elastic.co>

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.plan.logical.CommandLicenseTests testLicenseCheck #137168

MattAlp committed:
Catch-and-rethrow TooComplexToDeterminizeException within ESQL (#137024)

---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

fang-xing-esql committed:
[ES|QL] Non-Correlated Subquery in FROM command (#135744)

* non-correlated subquery in from command

leemthompo committed:
[docs] Update category field to use keyword type in full text tutorial (#137169)

closes https://github.com/elastic/docs-content/issues/3329

timgrein committed:
[Inference API] Rename E5 model/inference endpoint in EIS to JinaAI counterpart (#137028)

alex-spies committed:
ESQL: Add a bit of javadoc to preanalysis methods (#137099)

Explain where minimum transport version is determined for a given query.

kosabogi committed:
Adds ml-cpp release notes (#137178)

idegtiarenko committed:
Remove mutes pointing to closed ESQL flaky test issues (#137174)

pquentin committed:
rest-api-spec: Add missing options to enum types (#136640)

* rest-api-spec: Add missing options to enum types

* Allow 0-9 in enum options

leemthompo committed:
Add applies_to frontmatter to mapping-reference docs (#137184)

leemthompo committed:
[docs] Update ESQL command H1 titles for SEO (#137188)

Commit:
AnalyzerTest fix

Commit:
style fix

Commit:
Some fixes: change the TopNEncoder to the correct one, fix the unsupported test to match the expected results (good ones and not unsupported).

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Transport version

Commit:
[CI] Auto commit changes from spotless

Commit:
Remove temporary print

alex-spies reviewed (approved):
Ok, I see that the new type should be used in AllSupportedFieldsTestCase (original comment [here](https://github.com/elastic/elasticsearch/pull/133309#discussion_r2445527125)), so I consider this unblocked from my side :)

Let's not forget to attach the `test-release` label before merging!

craigtaverner commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:
```diff
@@ -12,6 +12,7 @@ setup:
           mappings:
             properties:
               aggregate_metric_double:
+
```
Why add this line?

shmuelhanoch replied:
Fixed

Commit:
Add missing fix in boolean tranport flag in IndexResolver

Commit:
Merge branch 'main' into esql_date_range_data_type

craigtaverner committed:
Serialize "unsupported" DataType when communicating to old clusters

craigtaverner committed:
Merge branch 'esql_date_range_data_type' of github.com:shmuelhanoch/elasticsearch into esql_date_range_data_type

craigtaverner committed:
Block date_range on yaml tests for bwc

mamazzol committed:
cat API: added endpoint for Circuit Breakers (#136890)

Added CAT Action to display Circuit Breakers stats for all nodes. The
API supports pattern matching as a path parameter and the standard query
parameters of CAT actions. This change includes spec and yamlRestTest.

Addresses #132688

nik9000 committed:
ESQL: Fix release tests (#137298)

New field type isn't serializable outside of snapshot.

MattAlp committed:
Reject invalid `reverse_nested` aggs (#137047)

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner committed:
Fixed missing import after resolving conflicts with main

craigtaverner committed:
Fix mistake on merging main with new element types

elasticsearchmachine committed:
Mute org.elasticsearch.xpack.esql.plan.physical.ShowExecSerializationTests testConcurrentSerialization #137338

elasticsearchmachine committed:
Mute org.elasticsearch.readiness.ReadinessClusterIT testReadinessDuringRestartsNormalOrder #136955

leemthompo committed:
[docs] Update changelog summary for semantic_text ELSER on EIS default (#137339)

mosche committed:
Fix method visibility in public-callers-finder (#137200)

Public-callers-finder did not correctly report if methods are externally
visible due to the following issues: - checking if interfaces are
exported (see findAccessibility) must be done using the exports of the
interface’s module, not the exports of the current class - super classes
must be treated the same way as interfaces, currently these are ignored
- currently all public methods of a (potentially private, non-exported)
class implementing a public, exported interface are considered
accessible / visible regardless if part of the interface or not

This fixes visibility to be consistent with the JdkApiExtractor tool by
implementing both using the same common logic in
`AccessibleJdkMethods.loadAccessibleMethods`.

Note: this currently includes
https://github.com/elastic/elasticsearch/pull/137193, which will be
merged independently.

Relates to ES-13117

jonathan-buttner committed:
[ML] Adding bulk create functionality to ModelRegistry (#136569)

* Adding bulk storage of multiple models

* Adding tests

* Adding log for duplicate ids

* [CI] Auto commit changes from spotless

* Removing unused code

* Removing constructor

* Adding more tests

* Adding in logic to delete models when a failure occurs

* revert rename changes

* formatting

* Starting on feedback

* Improving tests

* Moving most tests to ModelRegistryIT

* [CI] Auto commit changes from spotless

* Fixing test

* Removing duplicate tests

* Handling empty list and duplicates

* Fixing empty delete

---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

mridula-s109 committed:
Remove deprecated constant for default elser inference id (#137329)

* Cleaned up the old constant

* Cleaned up test

* Update docs/changelog/137329.yaml

* Update docs/changelog/137329.yaml

* Update docs/changelog/137329.yaml

* Delete docs/changelog/137329.yaml

craigtaverner committed:
Revert fix to DataType serialization

We should handle this at a higher level, since this code is widely used in serialization tests which fail with this change.

nik9000 committed:
ESQL: Handle release of 9.2 in test (#137070)

I'd made a mistake in #136327 when writing the test for fetching fields
that the release of 9.2.0 revealed. This fixes it and adds one more test
that we needed from #136327.

Kubik42 committed:
Fixed inconsistency in the isSyntheticSourceEnabled flag (#137297)

* Fixed inconsistency in the isSyntheticSourceEnabled flag

* Update docs/changelog/137297.yaml

prwhelan committed:
[ML] Disable CPS for Dataframes (#136716)

Cross-Project Search and Cross-Cluster Search is indefinitely disabled
for Dataframe Analytics. The error message will now display if the
syntax would have otherwise resolved to the respective feature.

mridula-s109 committed:
[Docs] Improve semantic_text updates documentation organization (#137340)

* Move script update restrictions section for consistency

* Update docs/reference/elasticsearch/mapping-reference/semantic-text.md

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

* Update docs/reference/elasticsearch/mapping-reference/semantic-text.md

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

* Update docs/reference/elasticsearch/mapping-reference/semantic-text.md

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

---------

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

romseygeek committed:
Remove unused field from IndexModule (#137342)

masseyke committed:
Improving random sampling performance by lazily calling getSamplingConfiguration() (#137223)

prwhelan committed:
[ML] Disable CrossProject for Datafeeds (#136897)

Initially, Datafeeds will not support cross-project source indices. We
will verify that the IndicesOptions is not trying to resolve a
cross-project index expression and throw a cross-project specific error
message.

neilbhavsar committed:
[ES-12998] Invoking gradle continue flag on periodic runs to allow for more information on test failures (#136900)

* Add --continue flag to invoke maven when a task failure has been hit so that we can see the outcome of more testing over time.

ioanatia committed:
ES|QL: Improve value loading for match_only_text mapping (#137026)

prwhelan committed:
[Transform] Remove extra reset calls (#137346)

ESRestTestCase already calls the `_reset` API, we do not need to do it
twice.

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

kderusso committed:
Add chunk_rescorer usage to output of explain and profile for text_similarity_rank_retriever (#137249)

* Add chunk_rescorer usage to output of explain for text_similarity_rank_retriever

* Update docs/changelog/137249.yaml

* Update toString

* Add support for profile

* Update 137249.yaml

* That's what you get for resolving conflicts in the UI, fixed compile failure

* [CI] Auto commit changes from spotless

* Fix test compilation

---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

kkrik-es committed:
Remove `first` and `last` functions from documentation (#137341)

* Remove `first` and `last` functions from documentation

* Apply suggestions from code review

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

* update

---------

Co-authored-by: Liam Thompson <leemthompo@gmail.com>

nik9000 committed:
ESQL: Work around concurrent serialization bug (#137350)

The bug still exists in the underlying code, but it isn't released so
I'm working around it by doing an eager call to `hashCode`. We'll fix
the real bug real soon.

Closes #137338

kderusso committed:
[ES|QL] Add CHUNK function (#134320)

* Add new function to chunk strings

* Refactor CHUNK function to support multiple values

* Default to returning all chunks

* [CI] Auto commit changes from spotless

* Handle warnings

* Loosen export restrictions to try to get compile error working

* Remove inference dependencies

* Fix compilation errors

* Remove more inference deps

* Fix compile errors from merge

* Fix existing tests

* Exclude from CSV tests

* Add more tests

* Cleanup

* [CI] Auto commit changes from spotless

* Cleanup

* Update docs/changelog/134320.yaml

* PR feedback

* Remove null field constraint

* [CI] Auto commit changes from spotless

* PR feedback: Refactor to use an options map

* Cleanup

* Regenerate docs

* Add test on a concatenated field

* Add multivalued field test

* Don't hardcode strings

* [CI] Auto commit changes from spotless

* PR feedback

---------

Co-authored-by: elasticsearchmachine <infra-root+elasticsearchmachine@elastic.co>

Commit:
Try a temporary fix for enrichment

Commit:
[CI] Auto commit changes from spotless

jordan-powers committed:
Add default sort for message.template_id field in logsdb indices (#136571)

This PR adds the index setting 
`index.logsdb.default_sort_on_message_template`. When set, LogsDB indices
will add `message.template_id` to the default sort fields (assuming the
`message` field is present and of type `pattern_text`).

Commit:
Some small fixes

Commit:
Transport version

Commit:
[CI] Update transport version definitions

Commit:
Specifically allow to_date_range and to_string

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Tansport version

shmuelhanoch committed:
Update docs/changelog/133309.yaml

shmuelhanoch committed:
Merge branch 'main' into esql_date_range_data_type

Commit:
Fixed forbidden toString

Commit:
Revert enrichment hack

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
transport version

Commit:
Fixed accidentally deleted files

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Generated files for to_string / is_null tests

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Transport version

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Transport version

Commit:
Small merge accident fix

Commit:
Added missing case for BlockTestUtils

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Small fix for AllSupportedFieldTest

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Transport version

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Transport version

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
merge fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
change date_range to under_construction

Commit:
Some error tests fixes

Commit:
More error tests fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
More merge fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Auto commit changes from spotless

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'esql_date_range_data_type' of github.com:shmuelhanoch/elasticsearch into esql_date_range_data_type

Commit:
Smale checstyle fix

Commit:
error tests fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
More fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
One more error fix

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Fix the date range capability

shmuelhanoch committed:
Update docs/changelog/133309.yaml

Commit:
bring back accidentally deleted files

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'esql_date_range_data_type' of github.com:shmuelhanoch/elasticsearch into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
changelog type

Commit:
AllSupportedFieldTest fix for no snapshot build

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Transport version

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Fix for CSV test to avoid order false errors

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Another fix after changing to under construction

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
merge fixes

Commit:
[CI] Auto commit changes from spotless

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/RangeFieldMapper.java:
```diff
@@ -61,6 +65,8 @@ public class RangeFieldMapper extends FieldMapper {
     public static final boolean DEFAULT_INCLUDE_UPPER = true;
     public static final boolean DEFAULT_INCLUDE_LOWER = true;
 
+    public static final TransportVersion ESQL_LONG_RANGES = TransportVersion.fromName("esql_long_ranges");
```
Alternatively we can make two copies of this in compute and esql-core, similarly to how `ESQL_AGGREGATE_METRIC_DOUBLE_CREATED_VERSION` is done.

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

craigtaverner commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/LongRangeBlockBuilder.java:
```diff
@@ -0,0 +1,187 @@
+        return this;
+    }
+
+    public LongRangeBlockBuilder copyFrom(LongRangeBlock block, int pos) {
```
This is only used in tests. Should it really exist? I wonder about the value of tests that only test a method that is not used anywhere outside the tests?

craigtaverner commented on x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BlockBuilderCopyFromTests.java:
```diff
@@ -36,7 +36,7 @@ public static List<Object[]> params() {
             }
             for (boolean nullAllowed : new boolean[] { false, true }) {
-                if (e == ElementType.AGGREGATE_METRIC_DOUBLE) {
+                if (e == ElementType.AGGREGATE_METRIC_DOUBLE || e == ElementType.LONG_RANGE) {
```
I think elasticsearch does allow multi-values for date-range. Is this limit here because we have not yet implemented this support?

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/PreAnalyzer.java:
```diff
@@ -96,8 +99,16 @@ protected PreAnalysis doPreAnalyze(LogicalPlan plan) {
             if (fn.name().equalsIgnoreCase("to_aggregate_metric_double")) {
                 useAggregateMetricDoubleWhenNotSupported.set(true);
             }
+            if (List.of("to_date_range", "to_string").contains(fn.name().toLowerCase(Locale.ROOT))) {
```
I don't understand the point of this code. It seems to relate to the comment in IndexResolver:279, which seems a copy of the aggregate-metric-double comment related to older cluster nodes supporting that type without reporting that they do. Is that really true for date_range too?

craigtaverner committed:
Fix some date->long text replacements we missed

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -268,13 +276,18 @@ private void doResolveIndices(
      *                                       report that they support the type? This exists because some remotes <strong>do</strong>
      *                                       support {@code dense_vector} without reporting that they do. And, for a while, we used the
      *                                       query itself to opt into reading these fields.
+     * @param useDateRangeWhenNotSupported does the query itself force us to use {@code date_range} fields even if the remotes don't
```
The comment in AllSupportedFieldsTestCase seems to clarify things for AMD, and it does not sound appropriate for date_range to me: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/AllSupportedFieldsTestCase.java#L1027-L1033

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/date.csv-spec:
```diff
@@ -1618,3 +1618,48 @@ from employees
+dateRangeToString
+required_capability: date_range_field_type
+
+FROM date_ranges
```
```suggestion
FROM decades
```

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-date_ranges.json:
```diff
@@ -0,0 +1,13 @@
+{
```
This file duplicates mapping-decades.json, so can be deleted.

craigtaverner reviewed (changes requested):
Mostly looks quite good, but a few things need clarification, and I think some small fixes.

craigtaverner committed:
Removed duplicated test data `date_ranges.csv`

craigtaverner committed:
Merge branch 'esql_date_range_data_type' of github.com:shmuelhanoch/elasticsearch into esql_date_range_data_type

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner committed:
Update transport version after merging main

craigtaverner committed:
Nicer string literal

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner committed:
Update transport version after merging main

craigtaverner committed:
TO_DATERANGE should be snapshot-only

craigtaverner committed:
Fix failing tests after moving TO_DATERANGE to snapshot-only

craigtaverner committed:
Remove `useDateRangeWhenNotSupported`

since that seems only relevant to partial support across multiple versions

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into esql_date_range_data_type

craigtaverner reviewed (approved):
I've made a few updates now, and will mark this as approved to we can get it merged if CI goes green. We need to do a number of things in followup PRs including:
* Generate the ToString evaluator
* Get all mullti-value tests working without disabling
* Add circuit breakers to the block factory
* Support aggregations (count, present, absent and mv-)

craigtaverner committed:
Merge branch 'main' into esql_date_range_data_type

craigtaverner committed:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

craigtaverner committed:
Disabled TO_DATE_RANGE on release builds of testInlineCast

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Auto commit changes from spotless

shmuelhanoch committed:
Update docs/changelog/133309.yaml

elasticsearchmachine commented:
Hi @shmuelhanoch, I've updated the changelog YAML for you.

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
transport version

Commit:
merge fixes

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
another transport version

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Another try in fixing the multi cluster tests

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Update transport version definitions

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Fixed compilation error

Commit:
Fix thread leak in ContextIndexSearcherTests.testMaxClause

Add missing executor cleanup in finally block to prevent thread leaks
when the test creates a ThreadPoolExecutor.

Commit:
Add queryContainsIndices utility method to EsqlTestUtils

This new method checks if a given ESQL query contains any specified indices, facilitating special handling for queries using indices loaded into multiple clusters. The method is integrated into MultiClusterSpecIT to streamline the logic for determining when to use remote indices, improving code clarity and maintainability.

Commit:
Fix for the last review fix

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
[CI] Auto commit changes from spotless

craigtaverner committed:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

alex-spies committed:
Merge remote-tracking branch 'upstream/main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type

Commit:
Merge branch 'main' into esql_date_range_data_type
